asim: add randomized range generation

wenyihu6 opened the pull request:
This patch enables random range configuration to be generated.

TestRandomized can now take another setting parameter rangeGen (default: uniform
rangeGenType, uniform keySpaceGenType, empty weightedRand).

These generators are part of the framework fields which persist across
iterations. The numbers produced by the generator shape the distribution across
iterations.

- rangeKeyGenType: determines range generator type across iterations (default:
uniformGenerator, min = 1, max = 1000)

- keySpaceGenType: determines key space generator type across iterations
(default: uniformGenerator, min = 1000, max = 200000)

- weightedRand: if non-empty, enables weighted randomization for range
distribution

This provides three modes for range generation:
1. Default: currently set to uniform distribution
2. Random: randomly generates range distribution across stores
3. Weighted Randomization: enables weighted randomization for range distribution
if and only if given weightedRand is non-empty

Part of: https://github.com/cockroachdb/cockroach/issues/106311

Release note: None

wenyihu6 committed:
asim: add randomized range generation

This patch enables random range configurations to be generated.

TestRandomized can now take another setting parameter rangeGen (default: uniform
rangeGenType, uniform keySpaceGenType, empty weightedRand).

These generators are part of the framework fields which persist across
iterations. The numbers produced by the generator shape the distribution across
iterations.

- rangeKeyGenType: determines range generator type across iterations (default:
uniformGenerator, min = 1, max = 1000)

- keySpaceGenType: determines key space generator type across iterations
(default: uniformGenerator, min = 1000, max = 200000)

- weightedRand: if non-empty, enables weighted randomization for range
distribution

This provides three modes for range generation:
1. Default: currently set to uniform distribution
2. Random: randomly generates range distribution across stores
3. Weighted Randomization: enables weighted randomization for range distribution
if and only if given weightedRand is non-empty

Part of: cockroachdb#106311

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107354)
<!-- Reviewable:end -->

kvoli reviewed (commented):
Nice stuff! Left some comments - lmk if you have any questions.

> Reviewed 8 of 8 files at r1, 4 of 4 files at r2, 8 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-:-N_toQs94EuVDi6vHlPD:bpo2b6e)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`-- commits` line 25 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_tmPdR6M0lNKwB-hu4):*
Just a reminder to fill this out.
___
*[`pkg/kv/kvserver/asim/state/config_loader.go` line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_tkeb_BG1v21GpHyC0:-N_tkeb_BG1v21GpHyC1:b-8t5u5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/4156a0308b0e7c5b16cc128310c6ee144159ee52/pkg/kv/kvserver/asim/state/config_loader.go#L302)):*
> ```Go
> 				s.SetNodeLocality(node.NodeID(), locality)
> 				storesRequired := z.StoresPerNode
> 				if storesRequired < 1 {
> ```

Could we unify this with the above addition? It feels error prone to be handling default values in two places. Also consider adding a TODO for moving these defaults to generation and panicing instead.
___
*[`pkg/kv/kvserver/asim/tests/rand_framework.go` line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_tljHrD9eVHMzM1FsI:-N_tljHrD9eVHMzM1FsJ:b-s9pgci) ([raw file](https://github.com/cockroachdb/cockroach/blob/4156a0308b0e7c5b16cc128310c6ee144159ee52/pkg/kv/kvserver/asim/tests/rand_framework.go#L58)):*
> ```Go
> 
> func (f randTestingFramework) getCluster() gen.ClusterGen {
> 	if !f.s.randOptions["cluster"] {
> ```

Convert to using struct fields, like we discussed offline.
___
*[`pkg/kv/kvserver/asim/tests/rand_framework.go` line 237 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_tnYTICRFwwcCh_niy:-N_tnYTICRFwwcCh_niz:b-cdkj4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dec2f8510a9e275647f0796ad231a88c850d854/pkg/kv/kvserver/asim/tests/rand_framework.go#L237)):*
> ```Go
> )
> 
> func convertInt64ToInt(num int64) int {
> ```

I'd prefer to just panic if here rather than truncate.
___
*[`pkg/kv/kvserver/asim/tests/rand_gen.go` line 34 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_tnuUd04sd-C8l4KnE:-N_tnuUd04sd-C8l4KnF:b-bfig0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dec2f8510a9e275647f0796ad231a88c850d854/pkg/kv/kvserver/asim/tests/rand_gen.go#L34)):*
> ```Go
> }
> 
> func NewRandomizedBasicRanges(
> ```

There are quite a few constructor functions but I wonder how much value they are providing? The panic below seems good - but perhaps we could scale back the encapsulation, given the structs are exported anyway?
___
*[`pkg/kv/kvserver/asim/workload/workload.go` line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_toA_95CmxtkIO9FdA:-N_toA_95CmxtkIO9FdB:b-21na0d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dec2f8510a9e275647f0796ad231a88c850d854/pkg/kv/kvserver/asim/workload/workload.go#L194)):*
> ```Go
> }
> 
> func (g *uniformGenerator) Num() int64 {
> ```

From offline - it would be better to keep these as writeKey/readKey for now - then add a comment describing an intention to re-use `pkg/workload/(kv|ycsb)` key gens so we test with the same stuff.
___
*[`pkg/kv/kvserver/asim/tests/rand_test.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_tk29LBr_skxljdWK-:-N_tk29MDj1P0rIlR67W:b1qckgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/4156a0308b0e7c5b16cc128310c6ee144159ee52/pkg/kv/kvserver/asim/tests/rand_test.go#L30)):*
> ```Go
> 
> // TestRandomized is a randomized testing framework which validates an allocator
> // simulator by creating randomized configurations, exposing potential edge
> ```

nit: allocation simulation. This test is asserting on the expected behavior of allocation, not the simulator.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/asim/state/config_loader.go` line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_tkeb_BG1v21GpHyC0:-Na8_pM81UsgLmK1oY40:b46rm0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/4156a0308b0e7c5b16cc128310c6ee144159ee52/pkg/kv/kvserver/asim/state/config_loader.go#L302)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Could we unify this with the above addition? It feels error prone to be handling default values in two places. Also consider adding a TODO for moving these defaults to generation and panicing instead.
</blockquote></details>

Do you mean something like [this](https://github.com/cockroachdb/cockroach/compare/master...wenyihu6:cockroach:panic2?expand=1) or [this](https://github.com/cockroachdb/cockroach/compare/master...wenyihu6:cockroach:panic?expand=1)?


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/asim/tests/rand_framework.go` line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_tljHrD9eVHMzM1FsI:-Na8aHqaEnUCfABJPZtt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4156a0308b0e7c5b16cc128310c6ee144159ee52/pkg/kv/kvserver/asim/tests/rand_framework.go#L58)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Convert to using struct fields, like we discussed offline.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`-- commits` line 25 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_tmPdR6M0lNKwB-hu4):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Just a reminder to fill this out.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/asim/tests/rand_gen.go` line 34 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_tnuUd04sd-C8l4KnE:-Na8dnoM4f5JIC7s2dUg:b-nudpl8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dec2f8510a9e275647f0796ad231a88c850d854/pkg/kv/kvserver/asim/tests/rand_gen.go#L34)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

There are quite a few constructor functions but I wonder how much value they are providing? The panic below seems good - but perhaps we could scale back the encapsulation, given the structs are exported anyway?
</blockquote></details>

Agreed. Removed some constructors and added panic checks in Generate() instead.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/asim/workload/workload.go` line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_toA_95CmxtkIO9FdA:-Na8fcWm8Zp6-rE1i4UZ:bmi7zac) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dec2f8510a9e275647f0796ad231a88c850d854/pkg/kv/kvserver/asim/workload/workload.go#L194)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

From offline - it would be better to keep these as writeKey/readKey for now - then add a comment describing an intention to re-use `pkg/workload/(kv|ycsb)` key gens so we test with the same stuff.
</blockquote></details>

This wasn't so easy since writeKey/readKey are not exported. I'm considering to mark them as exported from pkg/kv/kvserver/asim/workload/workload.go since we will want to export them when re-using key gens from `pkg/workload/(kv|ycsb)`. Alternatively, we can re-create them here. Wdyt?


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> Reviewed 10 of 10 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107354)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/asim/state/config_loader.go` line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_tkeb_BG1v21GpHyC0:-NaCnWmb7POTm4QvyyB3:bhicbz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4156a0308b0e7c5b16cc128310c6ee144159ee52/pkg/kv/kvserver/asim/state/config_loader.go#L302)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Do you mean something like [this](https://github.com/cockroachdb/cockroach/compare/master...wenyihu6:cockroach:panic2?expand=1) or [this](https://github.com/cockroachdb/cockroach/compare/master...wenyihu6:cockroach:panic?expand=1)?
</blockquote></details>

I think both. It should panic, but then we also need to update the existing code.
___
*[`pkg/kv/kvserver/asim/workload/workload.go` line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_toA_95CmxtkIO9FdA:-NaCo2gu6ZoMlLfrpcge:b-j3wm7x) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dec2f8510a9e275647f0796ad231a88c850d854/pkg/kv/kvserver/asim/workload/workload.go#L194)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

This wasn't so easy since writeKey/readKey are not exported. I'm considering to mark them as exported from pkg/kv/kvserver/asim/workload/workload.go since we will want to export them when re-using key gens from `pkg/workload/(kv|ycsb)`. Alternatively, we can re-create them here. Wdyt?
</blockquote></details>

Let's go with the re-create here option for now.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
The first four commits are from #107075.

wenyihu6 commented:
*[`pkg/kv/kvserver/asim/workload/workload.go` line 167 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-NamfEzRAPfu4bScILT5:-NamfEzRAPfu4bScILT6:b-aza5nb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee6b12c770ebd759201768b1593255177a75f45b/pkg/kv/kvserver/asim/workload/workload.go#L167)):*
> ```Go
> // TODO(wenyihu6): Instead of duplicating the key generator logic in simulators,
> // we should directly reuse the code from the repo pkg/workload/(kv|ycsb) to
> // ensure consistent testing.
> ```

Added the comment here.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/asim/workload/workload.go` line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_toA_95CmxtkIO9FdA:-NamfHbs742eWunik_4l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dec2f8510a9e275647f0796ad231a88c850d854/pkg/kv/kvserver/asim/workload/workload.go#L194)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Let's go with the re-create here option for now.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
Just a small question re: key generator in the `tests` pkg. Looking great!

> Reviewed 13 of 13 files at r7, 2 of 2 files at r8, 6 of 6 files at r9, 6 of 6 files at r10, 8 of 8 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-:-NaqeFZj5nJu9l4k29Nt:b-8wa84i)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`-- commits` line 65 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-NaqZTKP3oP5UB90JTEs):*
nit: "enables random range configurations to be generated"
___
*[`pkg/kv/kvserver/asim/tests/rand_framework.go` line 237 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_tnYTICRFwwcCh_niy:-NaqdYi08CRJoRJUAMoD:b-wdj8zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dec2f8510a9e275647f0796ad231a88c850d854/pkg/kv/kvserver/asim/tests/rand_framework.go#L237)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

I'd prefer to just panic if here rather than truncate.
</blockquote></details>

Nice, thanks!
___
*[`pkg/kv/kvserver/asim/tests/rand_gen.go` line 89 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-NaqdrfM3ht3Esfb72Ko:-NaqdrfM3ht3Esfb72Kp:b-2r6b5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee6b12c770ebd759201768b1593255177a75f45b/pkg/kv/kvserver/asim/tests/rand_gen.go#L89)):*
> ```Go
> // newUniformKeyGen returns a generator that generates number∈[min, max] with a
> // uniform distribution.
> func newUniformKeyGen(min, max int64, rand *rand.Rand) generator {
> ```

What is the difference between this key gen, and the `asim/workload` key gen?
___
*[`pkg/kv/kvserver/asim/workload/workload.go` line 167 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-NamfEzRAPfu4bScILT5:-NaqeEU65_V6A81bf3wH:b2nhb2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee6b12c770ebd759201768b1593255177a75f45b/pkg/kv/kvserver/asim/workload/workload.go#L167)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Added the comment here.
</blockquote></details>

Ack


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/asim/tests/rand_framework.go` line 237 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_tnYTICRFwwcCh_niy:-NarwWfj3xVj301HXrih:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dec2f8510a9e275647f0796ad231a88c850d854/pkg/kv/kvserver/asim/tests/rand_framework.go#L237)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Nice, thanks!
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/asim/tests/rand_gen.go` line 89 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-NaqdrfM3ht3Esfb72Ko:-NarwYMw31p7XrSdxE7j:b-dkfjhb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee6b12c770ebd759201768b1593255177a75f45b/pkg/kv/kvserver/asim/tests/rand_gen.go#L89)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

What is the difference between this key gen, and the `asim/workload` key gen?
</blockquote></details>

I don't think so. I think we discussed in https://github.com/cockroachdb/cockroach/pull/107354#pullrequestreview-1545872816 that we want to re-create them. But I'm not sure if I'm interpreting the question correctly. 


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/asim/tests/rand_gen.go` line 89 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-NaqdrfM3ht3Esfb72Ko:-NarwzWE4IanO-_3m2f1:b6ohsos) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee6b12c770ebd759201768b1593255177a75f45b/pkg/kv/kvserver/asim/tests/rand_gen.go#L89)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

I don't think so. I think we discussed in https://github.com/cockroachdb/cockroach/pull/107354#pullrequestreview-1545872816 that we want to re-create them. But I'm not sure if I'm interpreting the question correctly. 
</blockquote></details>

I don't there are any differences*


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/asim/state/config_loader.go` line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-N_tkeb_BG1v21GpHyC0:-NamfJmT091sqlgRoBGI:b-gnikq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4156a0308b0e7c5b16cc128310c6ee144159ee52/pkg/kv/kvserver/asim/state/config_loader.go#L302)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

I think both. It should panic, but then we also need to update the existing code.
</blockquote></details>

Done in #107075


<!-- Sent from Reviewable.io -->

kvoli reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Nice, apologies for the delay!

> Reviewed 5 of 8 files at r13, 4 of 4 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107354#-:-Nc2gtUEAQqZ0dRRuFv0:b-ulb9t9)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @wenyihu6)



<!-- Sent from Reviewable.io -->
